Libraries/LibWeb/HTML/SharedWorker.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2025, Tim Flynn <trflynn89@ladybird.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibWeb/Bindings/SharedWorkerPrototype.h>
+#include <LibWeb/DOM/Document.h>
+#include <LibWeb/DOM/Event.h>
+#include <LibWeb/HTML/EventLoop/EventLoop.h>
+#include <LibWeb/HTML/EventNames.h>
+#include <LibWeb/HTML/MessageEvent.h>
+#include <LibWeb/HTML/MessagePort.h>
+#include <LibWeb/HTML/Scripting/Environments.h>
+#include <LibWeb/HTML/SharedWorker.h>
+#include <LibWeb/HTML/SharedWorkerGlobalScope.h>
+#include <LibWeb/HTML/Window.h>
+#include <LibWeb/HTML/Worker.h>
+#include <LibWeb/Platform/EventLoopPlugin.h>
+
+namespace Web::HTML {
+
+GC_DEFINE_ALLOCATOR(SharedWorker);
+
+// https://html.spec.whatwg.org/multipage/workers.html#dom-sharedworker
+WebIDL::ExceptionOr<GC::Ref<SharedWorker>> SharedWorker::construct_impl(JS::Realm& realm, String const& script_url, Variant<String, WorkerOptions>& options_value)
+{
+    // FIXME: 1. Let compliantScriptURL be the result of invoking the Get Trusted Type compliant string algorithm with
+    //           TrustedScriptURL, this's relevant global object, scriptURL, "SharedWorker constructor", and "script".
+    auto const& compliant_script_url = script_url;
+
+    // 2. If options is a DOMString, set options to a new WorkerOptions dictionary whose name member is set to the value
+    //    of options and whose other members are set to their default values.
+    auto options = options_value.visit(
+        [&](String& options) {
+            return WorkerOptions { .name = move(options) };
+        },
+        [&](WorkerOptions& options) {
+            return move(options);
```

Comment:
moving an lvalue reference?

I hope I didn't write the first revision of this code in another file :sweat_smile:

Reply:
hmm I usually let clangd tell me when RVO is being stepped on, but it's silent here 🤔

Reply:
FYI I verified that without `move`, a copy happens here